### PR TITLE
Add JWT/JWKS support

### PR DESCRIPTION
**Pull request by ro-tex**
This PR introduces the generation, signing, and verification of JWTs.  
It also adds a new endpoint which exposes a public version of our JWKS, so others can use it in order to verify the JWTs signed by us.

### CODE REVIEW

File: Makefile
Comment (MSevey) [inline Makefile:102]: You shouldn't need to include `test` here. That is just going to run the short tests twice.
Long tests are skipped unless it is a long test run, but short test shouldn't have any skip code in them

File: hash/hash.go
Comment (MSevey) [inline hash/hash.go:16]: Follow Up: there must be some linting that this go repo doesn't have. skyd complains if exported items don't have comments.

File: hash/hash.go
Comment (MSevey) [inline hash/hash.go:22]: nit: extra newline

File: hash/hash.go
Comment (MSevey) [inline hash/hash.go:32]: Follow Up: Docstring

File: hash/hash.go
Comment (MSevey) [inline hash/hash.go:85]: Follow Up: This method could use inline comments.

File: hash/hash.go
Comment (MSevey) [inline hash/hash.go:41]: REQ: needs testing of this file.

File: main.go
Comment (MSevey) [inline main.go:31]: Follow Up: We need to make sure that this gets follow up all the way through our documentation. This will need to be added to the `.env` file right?

Comment (ro-tex) [reply main.go:31]: It defaults to `/accounts/conf/jwks.json`. So, if you don't set anything in your `.env`, it will look for the file there.
I'll add some explanation to the README.

File: Makefile
Comment (ro-tex) [reply Makefile:102]: Oh, you're right. :)

File: api/handlers.go
Comment (ro-tex) [reply api/handlers.go:82]: Done.

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:16]: F/U: https://github.com/SkynetLabs/skynet-accounts/issues/64

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:85]: I am not sure what I can add in those inline comments. The code is extremely straightforward and self-documenting.
It extracts the configuration for the `argon2` hashing (explained in the docstring) by extracting each bit of it in a separate block.

Example:  
```go
	salt, err = base64.RawStdEncoding.DecodeString(parts[4])
	if err != nil {
		return nil, nil, nil, err
	}
	ac.SaltLength = uint32(len(salt))
```
I can add a comment to this code which says "extracts the password salt" but that seems redundant to a level at which it reduces the redability of the code rather than enchancing it.

If you feel strongly about those comments I will add them but in my opinion they will not be an improvement.

File: main.go
Comment (MSevey) [reply main.go:31]: 👍

File: README.md
Comment (MSevey) [inline README.md:29]: cc @firyx 
we will need to update ansible/lastpass for the server set up scripts. We can wait until it is merged but wanted to call your attention to it.

File: hash/hash.go
Comment (MSevey) [reply hash/hash.go:85]: it is redundant if you leave out the why 😄 
Tell me the story, take me on this journey of decoding the hash.

imo inline comments should explain why we are doing each step as well as what each step is. I agree that each individual step is pretty straightforward in terms of what it is doing, but reviewing the code, or coming back to it to potentially edit it in 6 months, I can't pull the context of why each step is needed or if there are additionally order of operation concerns. 

Inline comments should save me the mental energy of having to read the code to understand the why.  

Example:
```
if version != argon2.Version {
		return nil, nil, nil, ErrIncompatibleVersion
}
```
I know this is checking for the correct version. But are there other considerations to consider with different versions? Do we need to offer compatibility code in the future if the version changes?

Comment (MSevey) [reply hash/hash.go:85]: Being able to understand the why through inline comments also helps reviewers provide better feedback or potential alternatives because it is clear what the developer's thought process was when writing the code.

Comment (ro-tex) [reply hash/hash.go:85]: > Tell me the story, take me on this journey of decoding the hash.

Good point, you're absolutely right. Comments coming right up.

File: jwt/jwt.go
Comment (ChrisSchinnerl) [inline jwt/jwt.go:347]: Is this necessary? Does 
```go
var set jwk.Set
err = json.Unmarshal(b, &set)
```
not work?

File: jwt/jwt.go
Comment (ChrisSchinnerl) [inline jwt/jwt.go:115]: Seems like you never need the parsed result so I wouldn't bother parsing and returning it

**Review by ChrisSchinnerl — APPROVED** [review]

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:28]: We could. These are the default values used by Kratos.

The main advantage of the argon2 algo is that the hash size is not the only determining factor in determining the complexity of the hashing process. We want slow hashing, so people won't be able to brute force it but not so slow that it would be bad for us as well. This is controlled by the other 3 parameters.

16 bytes is the recommended size by argon2's authors but I don't see a problem with doubling that.

Links:
 - https://www.ory.sh/choose-recommended-argon2-parameters-password-hashing/
 - https://password-hashing.net/argon2-specs.pdf (see "Recommended parameters")

If we choose to change the length, all existing passwords won't be affected.

File: api/handlers.go
Comment (peterjan) [inline api/handlers.go:87]: I think you're off here with the timezones because 4 lines below you do `exp.UTC()`.
So it should probably be 
```
if exp.UTC().before(time.Now().UTC()) {
    // handle expired token
}
```
is that  possible?

File: hash/hash.go
Comment (peterjan) [reply hash/hash.go:60]: maybe something like `func(us UserSecret) Equals(other UserSecret) bool`?

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:87]: I agree about the magic number, I'll add a descriptive comment explaining what each part of the string is.

File: hash/hash_test.go
Comment (peterjan) [inline hash/hash_test.go:13]: This unit test is not covering any of the exit flows that aren't the happy flow.

Comment (peterjan) [reply hash/hash_test.go:13]: You definitely want to cover `ErrInvalidHash` and `ErrIncompatibleVersion` at least right?

File: jwt/jwt.go
Comment (peterjan) [inline jwt/jwt.go:299]: does this mean we are reading this file on every validation? Seems like something we might want to cache right?

File: jwt/jwt.go
Comment (peterjan) [inline jwt/jwt.go:350]: Is this log level severe enough? Seems like this is a pretty critical issue.

File: jwt/jwt.go
Comment (ro-tex) [reply jwt/jwt.go:347]: I couldn't make it work. The `jwk.NewSet()` is this:
```go
func NewSet() Set {
	return &set{}
}
```
So, while the `Set` interface is public, the returned type is unexported. There might be a way to do this without this function call but I'm not sure how.

File: jwt/jwt.go
Comment (peterjan) [reply jwt/jwt.go:299]: Oh I just realised that it is cached, I'm wondering though why we're not reading the key set on init. Something like that shouldn't be lazy loaded imo but rather prevent the server from coming up even. I guess this also goes hand in hand with my comment on the severity of the warning that we print when the file is not there. Shouldn't we simply exit if our accounts server doesn't have a keys file?

File: jwt/jwt.go
Comment (peterjan) [inline jwt/jwt.go:323]:
```suggestion
// the set or an error.
```

File: jwt/jwt.go
Comment (ro-tex) [reply jwt/jwt.go:115]: Oh, this is an artifact of splitting the code into several PRs. This function is not actually used in this PR and should have been in the next one where you can find this line where we use the token to grab its expiration time:
https://github.com/SkynetLabs/skynet-accounts/pull/62/files#diff-855b4003387cd9e2b8e2c353c747fd1817eccf1d223b4066f0d7a1db0b4b4b4eR109

File: build/debug_off.go
Comment (ro-tex) [reply build/debug_off.go:2]: I am not sure. I know that `gofmt` adds the first one (`//go:build !debug`) if the second one is there and the second one is the old one which was used with go versions below 1.17.

So, unless we want to limit our users to go 1.17, I would prefer to keep both lines.

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:40]: Docstring updated and comments added.

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:43]: I think it would be very confusing for everyone reading this who's not really into the matter. Password hashing is a familiar concept to most people and throwing the word "key" at them would only create confusion. I, myself, don't know how argon2 works under the hood and I haven't looked into how this "key" is generated and used.

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:60]: Wait... These two things are not two of a kind. One of them is a plaintext secret coming from the user and the other is the hashed version of a secret that we store in the database. That's why I've chosen the variable names there.

I was going to respond to https://github.com/SkynetLabs/skynet-accounts/pull/61#discussion_r700884476 directly but this is a very good example of the problem that would create. `UserSecret` does not tell us what we have in there - whether it's a plaintext, a ciphertext, or a hash. Or a "key" in the case of argon2, which functions like a hash for our purposes. I agree that neither does `[]byte` but that at least doesn't add to the confusion. 

I propose to keep referring to argon2 keys as hashes because we use them as such and it will add less confusion.  
I will add a type called `Argon2HashRecord` for the `[]byte` returned by argon2.  
I will change the `Compare` function to accept a string and a Argon2HashRecord types, so it's clearer what should go where.

Does that sound like a good solution?

File: hash/hash.go
Comment (ro-tex) [reply hash/hash.go:56]: Added type `Argon2HashRecord`.

File: api/handlers.go
Comment (ro-tex) [reply api/handlers.go:87]: Yes, well spotted! Thanks!

File: hash/hash_test.go
Comment (ro-tex) [reply hash/hash_test.go:13]: I added this quickly just to ensure the method does what I expect it to. And then I forgot to expand it. :)
More cases added.

File: jwt/jwt.go
Comment (ro-tex) [reply jwt/jwt.go:299]: Yes, that makes sense. While writing this I was thinking about the mixed case where we're running with both accounts and Kratos signing JWTs but it looks more and more likely that once we deliver this we will immediately axe Kratos. I will make it required.

File: jwt/jwt.go
Comment (ChrisSchinnerl) [reply jwt/jwt.go]: ohhh it's an interface. nevermind then

**Review by MSevey — APPROVED** [review]

**Review by ChrisSchinnerl — APPROVED** [review]

**Pull request merged**
